man/serve.1
.\" Generated with Ronnjs 0.3.8
.\" http://github.com/kapouer/ronnjs/
.
.TH "YEOMAN\-SERVER" "1" "July 2012" "" ""
.
.SH "NAME"
\fByeoman-server\fR \-\- Launch a preview server
.
.SH "SYNOPSIS"
.
.nf
yeoman server
.
.fi
.
.SH "DESCRIPTION"
This command \.\.\.
.
.SH "SEE ALSO"
.
.IP "\(bu" 4
yeoman\-folders(1)
.
.IP "" 0



changelog.md
### 0.9.5 - 2012-11-05

- [#627](https://github.com/yeoman/yeoman/issues/627)
Fix: Usemin:html adds image file before a 'a' anchor

- [#636](https://github.com/yeoman/yeoman/issues/636)
Python3 support for Insight

- [#603](https://github.com/yeoman/yeoman/issues/603)
Add ability to specify a base path in the usemin block comment

- [#660](https://github.com/yeoman/yeoman/issues/660)
Determine linefeed from the content not the platform

- Upgrade to latest Bower


### 0.9.4 (12 October, 2012)

* Updated to Bower 0.2.0 (Issue [#551](https://github.com/yeoman/yeoman/issues/551))
* Less verbose `yeoman init` (Issue [#605](https://github.com/yeoman/yeoman/issues/605))
* Ability to disable Insight (Issue [#305](https://github.com/yeoman/yeoman/issues/385))
* Corrections to generator docs (Issue [#608](https://github.com/yeoman/yeoman/issues/608))
* Fixes to usemin:html skipping anchor tags with images as href (Issue [#615](https://github.com/yeoman/yeoman/issues/615))
* Fixes to main.css not being replaced with compiled version unless tag structure is exactly as yeoman expects it (Issue [#502](https://github.com/yeoman/yeoman/issues/502))
* Stops yeoman from looking for win32 binaries in ../vendor/ (Pull [#519](https://github.com/yeoman/yeoman/pull/519))
* Support for installing Yeoman behind a corporate proxy server (Pull [#587](https://github.com/yeoman/yeoman/pull/587))
* Remove a trailing comma in the options of the HTML task and add missing semicolons in the server task
(Issue [#589](https://github.com/yeoman/yeoman/pull/589))
* Fix to enable blank lines in usemin blocks (Issue [#560](https://github.com/yeoman/yeoman/issues/560))
* Updates to Insight documentation
* Fixes usemin:css inserting invalid <link> element (Issue [#586](https://github.com/yeoman/yeoman/issues/586))
* Fix for usemin replace task incorrectly matcheing filename when files in subfolders have the same name (Issue [#565](https://github.com/yeoman/yeoman/issues/565))
* New Windows [installation](https://github.com/yeoman/yeoman/wiki/Manual-Install) instructions 

### 0.9.3 (3 October, 2012)

* Temporary fix to allow Bower dependencies to be correctly copied to app/components. As a result of this patch RequireJS wiring of Bower deps is currently disabled

### 0.9.2 

TBU

### 0.9.1 (15 September, 2012)

* Removal of stdout checks, test against fs (Pull [#473](https://github.com/yeoman/yeoman/pull/473)).

* Fix for  initializer example (Pull [#447](https://github.com/yeoman/yeoman/pull/477 )).

* Support for an expanded compass --require arg (Pull [#483](https://github.com/yeoman/yeoman/pull/483)).   

* Yeoman setup gets a colored art (Pull [#493](https://github.com/yeoman/yeoman/pull/493))!

* Added ability to configure app index file for rjs task (Pull [#505](https://github.com/yeoman/yeoman/pull/505)).

* yeoman server:test and access to the app js files (Issue [#443](https://github.com/yeoman/yeoman/issues/443)).

* yeoman install backbone deletes existing files in app/scripts/vendor issue was fixed (Issue [#460](https://github.com/yeoman/yeoman/issues/460)).

* Replaced coffescript task with grunt-coffee (Pull[#522](https://github.com/yeoman/yeoman/pull/522))


cli/tasks/serve.js

var fs = require('fs'),
  path = require('path'),
  connect = require('connect'),
  socketio = require('socket.io'),
  open = require('open');

// client-side socket-io script
var ioScript = fs.readFileSync(path.join(__dirname, '../lib/support/reload.js'), 'utf8');

module.exports = function(grunt) {

  grunt.registerTask('server', 'Start a custom static web server.', function(target) {
    if(!grunt.config('server')) grunt.config('server', {
      staging: { port: 3000, base: 'intermediate/' },
      output: { port: 3001, base: 'publish/' }
    });

    if(target) this.requiresConfig('server.' + target);
    // long running process
    var cb = this.async();
    // Project configuration.
    var config = grunt.config('server');
    // if reload flag is setup, in server config or via `--reload` cli option
    // then starts a socket.io server instead
    var reload = config.reload || grunt.option('reload');
    // the helper name to use
    var helper = reload ? 'serve.io' : 'serve';

    // open the browser
    open('http://localhost:3000');

    // start the webserver(s)
    if(target) return grunt.helper(helper, config[target]);
    grunt.helper(helper, config.staging);
    grunt.helper(helper, config.output);
  });

  // **serve** creates and returns a basic application server, a configuration
  // object may be passed where:
  //
  // - base   - is the base directory to server static file from
  // - port   - server listen to this port
  grunt.registerHelper('server', function(config) {
    config = config || {};
    var base = config.base || './',
      port = config.port || 8000,
      app = connect();

    app.listen(port, function(err) {
      if(err) grunt.log.error(err).warn(err.message, err.code);
      grunt.log.ok('Started static web server in ' + base.underline.bold + ' on port ' + (port + '').green + '...');
      app.emit('start');
    });

    return app;
  });

  // **serve.io** creates and returns a webserver and setup a socket.io instance and
  // the "inject" middleware. `/socket.io/socket.io.js` and `/relod.js` are
  // then available.
  grunt.registerHelper('serve.io', function serve(config) {
    // start on given port / base
    var app = grunt.helper('server', config);

    app
      .use(grunt.helper('inject.io', config))
      .use(connect.static(config.base));

    // setup scoket.io
    var io = socketio.listen(app);
    io.set('log level', config.loglevel || 1);
    // and return socket.io server instance
    return io;
  });

  // **serve** same as serve.io, without the socket.io connection and inject
  // middleware.
  grunt.registerHelper('serve', function serve(config) {
    return grunt.helper('server', config)
      .use(connect.static(config.base));
  });

  // **inject.io** is a grunt helper returning a valid connect / express middleware.
  // It's job is to setup a middleware right before the usual static one, and to
  // bypass the response of `.html` file to render them with additional scripts.
  grunt.registerHelper('inject.io', function(config) {
    grunt.utils._.defaults(config, {
      hostname: 'localhost'
    });

    return function inject(req, res, next) {
      // build filepath from req.url and deal with index files for trailing `/`
      var filepath = req.url.slice(-1) === '/' ? req.url + 'index.html' : req.url;

      // the client-side template
      var template = grunt.utils._.template(ioScript);

      // deal with our special socket.io client-side script
      if(path.basename(filepath) === 'reload.js') {
        res.setHeader('Content-Type', connect.static.mime.lookup('js'));
        return res.end(template(config));
      }

      // if ext is anything but .html, let it go through usual connect static
      // middleware
      if(path.extname(filepath) !== '.html') return next();

      // setup some basic headers, might add some.
      res.setHeader('Content-Type', connect.static.mime.lookup(filepath));

      // can't use the ideal stream / pipe case, we need to alter the html response
      // by injecting that little socket.io client-side app.
      filepath = path.join(config.base, filepath.replace(/^\//, ''));
      fs.readFile(filepath, function(e, body) {
        if(e) {
          res.writeHead(500);
          return res.end('Error loading' + req.url + '  -- ' + JSON.stringify(e));
        }

        body = body.toString().replace(/<\/script>/, function(w) {
          return [
            w,
            '  <script src="/socket.io/socket.io.js"></script>',
            '  <script src="/reload.js"></script>'
          ].join('\n');
        });

        res.end(body);
      });

      return next();
    };
  });

};
